Source/cryptography/implementation/OpenSSL/Vault.cpp

```diff
@@ -24,64 +24,156 @@
 #include <cryptalgo/cryptalgo.h>
 
 #include <openssl/evp.h>
+#include <openssl/rand.h>
+#include <openssl/sha.h>
+#include <openssl/hmac.h>
 
 #include "Vault.h"
 
 
 namespace Implementation {
 
-const uint32_t encryptionKeyId = 1;
-const uint32_t encryptionIVId = 2;
 
-/* static */ Vault& Vault::Instance()
+static constexpr uint8_t IV_SIZE = 16;
+
+namespace Netflix {
+
+static constexpr uint32_t KPE_ID = 1;
+static constexpr uint32_t KPH_ID = 2;
+static constexpr uint32_t KPW_ID = 3;
+static constexpr uint32_t ESN_ID = 4;
+
+uint16_t HKDF(const uint16_t inSize, const uint8_t input[], const uint16_t in2Size, const uint8_t input2[], uint8_t output[])
 {
-    static Vault instance;
+    // As per https://github.com/Netflix/msl/wiki/Pre-shared-Keys-or-Model-Group-Keys-Entity-Authentication
+
+    uint8_t inputBuffer[inSize + in2Size];
+    ::memcpy(inputBuffer, input, inSize);
+    ::memcpy(inputBuffer + inSize, input2, in2Size);
+
+    static const uint8_t salt[] = { 0x02, 0x76, 0x17, 0x98, 0x4f, 0x62, 0x27, 0x53, 0x9a, 0x63, 0x0b, 0x89, 0x7c, 0x01, 0x7d, 0x69 };
+    static const uint8_t data[] = { 0x80, 0x9f, 0x82, 0xa7, 0xad, 0xdf, 0x54, 0x8d, 0x3e, 0xa9, 0xdd, 0x06, 0x7f, 0xf9, 0xbb, 0x91 };
+
+    uint8_t hmac[SHA256_DIGEST_LENGTH];
+    uint32_t hmacSize = 0;
+    HMAC(EVP_sha256(), salt, sizeof(salt), inputBuffer, sizeof(inputBuffer), hmac, &hmacSize);
+    ASSERT(hmacSize == SHA256_DIGEST_LENGTH);
+
+    uint32_t outSize = 0;
+    HMAC(EVP_sha256(), hmac, hmacSize, data, sizeof(data), output, &outSize);
+    ASSERT(outSize == SHA256_DIGEST_LENGTH);
+
+    return (outSize);
+}
+
+} // namespace Netflix
+
+
+/* static */ Vault& Vault::NetflixInstance()
+{
+    static Vault instance(CRYPTOGRAPHY_VAULT_NETFLIX);
     return (instance);
 }
 
-Vault::Vault()
+Vault::Vault(cryptographyvault id)
     : _lock()
     , _items()
-    , _lastHandle(0x80000000)
+    , _lastHandle(0)
+    , _id(id)
+    , _vaultKey({ 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff, 0x11 }, 16)
 {
-    typedef uint8_t pkey[16];
+    /* Netflix vault specific initialisation */
+    if (id == CRYPTOGRAPHY_VAULT_NETFLIX) {
+        std::string path;
+        WPEFramework::Core::SystemInfo::GetEnvironment(_T("NETFLIX_VAULT"), path);
+        WPEFramework::Core::File file(path.c_str(), true);
+
+        if (file.Open(true) == true) {
+            struct NetflixData {
```

Comment:
Pragma pack(1) might be usefull on this as we do not know who or where it is compiled

Reply:
It's there!

https://github.com/rdkcentral/Thunder/blob/e10166410a1a5d8ae09b337671d3e19684f261da/Source/cryptography/implementation/OpenSSL/Vault.cpp#L59-L65